pkg/southbound/synchronizer/sessionManager.go

```diff
@@ -0,0 +1,246 @@
+	}
+}
+
+// WithSouthboundErrChan sets southbound error channel
+func WithSouthboundErrChan(southboundErrorChan chan events.DeviceResponse) func(*SessionManager) {
+	return func(factory *SessionManager) {
+		factory.southboundErrorChan = southboundErrorChan
+	}
+}
+
+// WithDispatcher sets dispatcher
+func WithDispatcher(dispatcher *dispatcher.Dispatcher) func(*SessionManager) {
+	return func(sessionManager *SessionManager) {
+		sessionManager.dispatcher = dispatcher
+	}
+}
+
+// WithModelRegistry sets model registry
+func WithModelRegistry(modelRegistry *modelregistry.ModelRegistry) func(*SessionManager) {
+	return func(sessionManager *SessionManager) {
+		sessionManager.modelRegistry = modelRegistry
+	}
+}
+
+// WithOperationalStateCache sets operational state cache
+func WithOperationalStateCache(operationalStateCache map[topodevice.ID]devicechange.TypedValueMap) func(*SessionManager) {
+	return func(sessionManager *SessionManager) {
+		sessionManager.operationalStateCache = operationalStateCache
+	}
+}
+
+// WithNewTargetFn sets southbound target function
+func WithNewTargetFn(newTargetFn func() southbound.TargetIf) func(*SessionManager) {
+	return func(sessionManager *SessionManager) {
+		sessionManager.newTargetFn = newTargetFn
+	}
+}
+
+// WithOperationalStateCacheLock sets operational state cache lock
+func WithOperationalStateCacheLock(operationalStateCacheLock *sync.RWMutex) func(*SessionManager) {
+	return func(sessionManager *SessionManager) {
+		sessionManager.operationalStateCacheLock = operationalStateCacheLock
+	}
+}
+
+// WithDeviceChangeStore sets device change store
+func WithDeviceChangeStore(deviceChangeStore device.Store) func(*SessionManager) {
+	return func(sessionManager *SessionManager) {
+		sessionManager.deviceChangeStore = deviceChangeStore
+	}
+}
+
+// Start starts session manager
+func (sm *SessionManager) Start() error {
+	log.Info("Session manager started")
+	err := sm.deviceStore.Watch(sm.topoChannel)
+	if err != nil {
+		return err
+	}
+
+	go sm.processDeviceEvents(sm.topoChannel)
+	return nil
+}
+
+// processDeviceEvents process incoming device events
+func (sm *SessionManager) processDeviceEvents(ch <-chan *topodevice.ListResponse) {
+	for event := range ch {
+		log.Infof("Received event type %v for device %v", event.Type, event.Device)
+		err := sm.processDeviceEvent(event)
+		if err != nil {
+			log.Errorf("Error updating session %v", event.Device.ID, err)
+		}
+
+	}
+}
+
+// processDeviceEvent process a device event
+func (sm *SessionManager) processDeviceEvent(event *topodevice.ListResponse) error {
+
+	switch event.Type {
+	case topodevice.ListResponse_ADDED:
+		log.Info("Process device event Added")
+		err := sm.createSession(event.Device)
+		if err != nil {
+			return err
+		}
+
+	case topodevice.ListResponse_NONE:
+		err := sm.createSession(event.Device)
+		if err != nil {
+			return err
+		}
+	case topodevice.ListResponse_UPDATED:
+		log.Info("Process device event updated")
```

Comment:
@kuujo 
In the current code, we check the address and if there is a change, we reconnect but if we change any other fields, we also need to update the state of a device. Should we create a new session anyway?

Reply:
I think either way is fine.